### ISSUE TITLE
Manage circular dependencies by modelling initial resource state

### ISSUE THREAD

**Issue by danielrbradley** [pulumi/pulumi#13999]
This is a proposal for what should be a concrete and achievable resolution to the long standing issue of circular dependencies:
- https://github.com/pulumi/pulumi/issues/3021

### Implemention

- Add a SDK resource option called: `isInitialState: bool`. This indicates that this register resource call is special and expects a final state to be declared subsequently.
- The initial and final registered resources should be their own entries in the state file.
- We might need a variation in the URN of the initial resource to ensure it doesn't conflict with the final (normal) state URN.
- Any resource linking back to an initial state which would cause a replacement would be a hard failure where the engine would not continue. To perform a replacement, the initial resource would first have to recreate itself, then the second would be an update to get back to the final state having also updated any dependant resource.
- Add an SDK resource option for `skipRevertOnDelete: bool`. If set to `true` this will just wait for depdendencies on the initial state to be deleted, then delete this resource at the point of the initial resource being delete from the graph. If not set, the default behaviour is for the state to be reverted back to the initial state when deleting the final state, then delete dependencies of the initial state, then actually delete the resource at the point of deleting the initial resource in the graph.
- Add an SDK resource option for `deleteBeforeDependencies: bool`. When set to true, this will attempt to delete the resource immediately at the point of deleting the final resource state, then delete dependencies, then do no action when deleting the initial state in the graph.

#### Initial program

Program:
```js
const initalA = new A("my-a", {}, {
  isInitialState: true,
});

const b = new B("my-b", { propA: initialA.name });

const finalA = new A("my-a", { propB: b.name });
```

First preview:

```
     Type                                 Name         Plan       
     pulumi:pulumi:Stack       my-stack            
 +   └─ provider:ns:A         my-a          create
 +   └─ provider:ns:B         my-b          create
 ~   └─ provider:ns:A         my-a          update     [diff: ~properties]
```

Updating an unrelated property after first deployment:

```js
const initalA = new A("my-a", {
  otherProp: "foo",
}, {
  isInitialState: true,
});

const b = new B("my-b", { propA: initialA.name });

const finalA = new A("my-a", {
  otherProp: "bar",
  propB: b.name,
});
```

The initial property change would not be shown

Update preview:

```
     Type                                 Name         Plan       
     pulumi:pulumi:Stack       my-stack            
 ~   └─ provider:ns:A         my-a          update     [diff: ~otherProp]
```

On causing a replacement - the replacement would have to affect the initial resource first - then match in the subsequent update:

```js
const initalA = new A("my-a", {
  sku: "different",
}, {
  isInitialState: true,
});

const b = new B("my-b", { propA: initialA.name });

const finalA = new A("my-a", {
  sku: "different",
  propB: b.name,
});
```

Preview:

```
     Type                                 Name         Plan       
     pulumi:pulumi:Stack       my-stack            
 ~   └─ provider:ns:A         my-a          replace    [diff: ~sku]
 ~   └─ provider:ns:A         my-a          update     [diff: ~properties]
```

### Variations

- The link back to the initial resource is implicit based on them having the same name, but could instead be a resource option of `initialResource: Resource`.
- We could perhaps use the same URN for both the initial and final states, but it would need investigation what else this might break.

### Questions

- What if resource options change between calls? We could assert that all key options are exactly the same (full list of behaviour still to be defined). Dependencies would be allowed to vary. Everything else should likely be forced to be identical. This should be an error if not the same.

**Comment by Mondonno** [comment]
C'mon this was almost 2 years ago and nothing has been done about that? Cyclic dependencies are very annoying currently in Pulumi.

**Comment by ashish111333** [comment]
@danielrbradley  can I open a draft PR on this ?

**Comment by ashish111333** [comment]
@danielrbradley  is this proposal still valid ?  @Frassle  can I work on this ?

**Comment by Frassle** [comment]
The problem is still a problem, I'm not sure the proposal as is will be right to solve it. It's a hard problem. But feel free to play with ideas here, just will be a challenge to get something that's actually shippable in product.

**Comment by ashish111333** [comment]
@Frassle  agreed , is a lot of work but I can open a draft PR as POC or should I start with a Detailed design doc and get it approved first  ?
